repair/row_level.cc

```diff
@@ -3166,6 +3223,10 @@ repair_service::insert_repair_meta(
                 reason,
                 compaction_time] (reader_permit permit) mutable {
         node_repair_meta_id id{from, repair_meta_id};
+        sstables::run_id run_id;
+        auto [it, _] = repair_ranges_map().try_emplace(from, range_tracker{});
+        it->second.insert_range(s->id(), range);
```

Comment:
So you handle it when RF > 1, by having one range tracker for each source.

What happens if there's parallel repair across tables?

So we cannot really rely on run ID, right? as they could be bogus.

Note that with RF > 1, reshape is needed anyway to merge data from different sources.

Reply:
If we want to have the state per table, we can move it to table and clear it as soon as maintenance set is cleared.

Reply:
If I have overlapping produced by RF > 1, I still need to perform reshape so the run id wouldn't be helpful, unless we forgive the need for reshape and allow regular to catch it up, but then we wouldn't be avoiding additional write amplification.

Reply:
> So you handle it when RF > 1, by having one range tracker for each source.

Yes

> 
> What happens if there's parallel repair across tables?

I think it can happen today using the api but we don't do that.
I mentioned it in the change log.

> 
> So we cannot really rely on run ID, right? as they could be bogus.

In theory.  In practice it should be fine.

> Note that with RF > 1, reshape is needed anyway to merge data from different sources.

That's fine.  With the run_id we can do and efficient incremental reshape compaction.

> If we want to have the state per table, we can move it to table and clear it as soon as maintenance set is cleared.

Can you please point at when repair is or can be notified that the above happened?

> If I have overlapping produced by RF > 1, I still need to perform reshape so the run id wouldn't be helpful, 

why?  wouldn't it help ICS reshape compaction?

> unless we forgive the need for reshape and allow regular to catch it up, but then we wouldn't be avoiding additional write amplification.

where does it come from?

Reply:
> > So you handle it when RF > 1, by having one range tracker for each source.
> 
> Yes
> 
> > What happens if there's parallel repair across tables?
> 
> I think it can happen today using the api but we don't do that. I mentioned it in the change log.
> 
> > So we cannot really rely on run ID, right? as they could be bogus.
> 
> In theory. In practice it should be fine.
> 

Today, users can have regular repair triggered on all tables of a keyspace simulteanously. You're thinking about the one-table-at-a-time that applies to repair-based operations, I was thinking about regular repair where there's not really an enforcement. 

But perhaps that's fine. The worst case scenario we picked different run IDs, and it's no different than today.

> > Note that with RF > 1, reshape is needed anyway to merge data from different sources.
> 
> That's fine. With the run_id we can do and efficient incremental reshape compaction.

Incremental compaction doesn't really depend on run id, but rather token range of individual sstables. The run id would only be interesting to avoid bypassing the entire reshape and forwarding the run as is for main set, no?

> 
> > If we want to have the state per table, we can move it to table and clear it as soon as maintenance set is cleared.
> 
> Can you please point at when repair is or can be notified that the above happened?
> 

If range tracker is tied with the lifetime of a repair job (is it?), it seems to me it won't work with regular repair for the reason I explained before. SM deliberately splits a table repair into multiple jobs, one for each range.